src/learner.cc

```diff
@@ -406,8 +407,29 @@ class LearnerConfiguration : public Learner {
   }
 
   void LoadConfig(Json const& in) override {
+    // If configuration is loaded, ensure that the model came from the same version
+    // For now, enforce this requirement for R only; throw warning for Python users
     CHECK(IsA<Object>(in));
-    Version::Load(in);
+    auto origin_version = Version::Load(in);
+
+    if (!Version::Same(origin_version)) {
+#if XGBOOST_STRICT_R_MODE
+      LOG(FATAL) << "You are using XGBoost " << Version::String(Version::Self())
+                 << " and attempting to load model from XGBoost "
+                 << Version::String(origin_version) << " via readRDS()."
+                 << "It is not possible to load an old model from RDS; please use "
+                 << "xgb.save() and xgb.load() for long-term archival of models. "
+                 << "For more information, visit "
+                 << "https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html";
+#else  // XGBOOST_STRICT_R_MODE
+      LOG(FATAL) << "You are using XGBoost " << Version::String(Version::Self())
+                 << " and attempting to load model from XGBoost "
+                 << Version::String(origin_version) << " via pickle.load() or a similar "
+                 << "mechanism. For long-term archival of models, use the save_model() method. "
+                 << "For more information, visit "
+                 << "https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html";
+#endif  // XGBOOST_STRICT_R_MODE
```

Comment:
@trivialfis  This PR will also disallow loading from old Python pickles. Can you review the error meesage?